Support for Active Directory multidomain forest.

nE0sIghT opened the pull request:
In AD forest samAccountName (or uid) may not be unique in the entire forest and userPrincipalName contains "@" symbol disallowed in Matrix User Identifiers.
This commit adds mxid generation logic to work around above restrictions.

Please note that ```exactly``` this commit was not yet tested by me with real AD forest, but very same implementation was tested.
Exactly this change will be tested by me next week.

richvdh commented on ldap_auth_provider.py:
```diff
@@ -88,6 +93,20 @@ def check_password(self, user_id, password):
             defer.returnValue(False)
         # user_id is of the form @foo:bar.com
         localpart = user_id.split(":", 1)[0][1:]
+        uid_value = localpart
+        default_givenName = localpart
```
please could you add some comments here? what do each of these variables do, and what is the difference between `localpart` and `uid_value` ?

richvdh commented on ldap_auth_provider.py:
```diff
@@ -88,6 +93,20 @@ def check_password(self, user_id, password):
+                if not self.ldap_default_domain:
+                    defer.returnValue(False)
+            else:
+                [login, domain] = localpart.rsplit(self.ldap_separator, 1)
```
a tuple is more conventional here:

```suggestion
                (login, domain) = localpart.rsplit(self.ldap_separator, 1)
```

richvdh commented on ldap_auth_provider.py:
```diff
@@ -88,6 +93,20 @@ def check_password(self, user_id, password):
+
+            if self.ldap_separator not in localpart:
+                if not self.ldap_default_domain:
+                    defer.returnValue(False)
```
it would probably be a good idea to log something here, otherwise people are going to be confused when their ldap logins fail silently.

richvdh commented on ldap_auth_provider.py:
```diff
@@ -88,6 +93,20 @@ def check_password(self, user_id, password):
             defer.returnValue(False)
         # user_id is of the form @foo:bar.com
         localpart = user_id.split(":", 1)[0][1:]
+        uid_value = localpart
+        default_givenName = localpart
+        if self.ldap_ad_forest:
+            login = localpart
+            domain = self.ldap_default_domain
+
+            if self.ldap_separator not in localpart:
+                if not self.ldap_default_domain:
+                    defer.returnValue(False)
+            else:
+                [login, domain] = localpart.rsplit(self.ldap_separator, 1)
+
+            uid_value = login + self.ldap_separator + domain
+            default_givenName = login
```
could you factor all this out to a separate method? I think it would be clearer that way.

richvdh commented on ldap_auth_provider.py:
```diff
@@ -245,6 +264,14 @@ def check_3pid_auth(self, medium, address, password):
+                [login, domain] = localpart.rsplit("@", 1)
+                localpart = login + self.ldap_separator + domain
+
+                if self.ldap_default_domain and domain == self.ldap_default_domain:
+                    user_id = self.account_handler.get_qualified_user_id(localpart)
+                    if (yield self.account_handler.check_user_exists(user_id)):
+                        localpart = login
```
why is the default domain special-cased here? why is it correct to strip the domain for accounts on that domain but not others?

nE0sIghT replied:
2 thoughts here:

1. This is very common pattern in corporate environment.  
You can look eg to [smb.conf's winbind use default domain](https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#idm10651).
Yes, it may be not recommended, bu it is widelly used.

2. Upgrading from previous versions of ```ldap_auth_provider``` while using ```samAccountName``` as uid.
After this PR merged it will be possible to continue using ```samAccountName``` as matrix uid for default domain users but also allow to use ```userPrincipalName``` for users of other domains in forest.

P.S.: ```samAccountName``` is first part of ```userPrincipalName``` before ```@```

richvdh commented on ldap_auth_provider.py:
```diff
@@ -347,6 +374,21 @@ class _LdapConfig(object):
+
+            ldap_config.default_domain = config.get("default_domain", None)
+            ldap_config.mxid_domain_separator = config.get("mxid_domain_separator", "/")
+            if ldap_config.mxid_domain_separator not in ("=", "/"):
+                raise Exception(
+                    "Only = or / symbols supported as domain separator"
+                )
```
why is this restriction needed?

nE0sIghT replied:
Those 2 symbols are only symbols allowed in mxid, but disallowed in ```userPrincipalName```.  
However let's drop this configuration in favor of binding ```/``` in mxid and using classic ```domain\login``` form as you suggested.

richvdh commented on ldap_auth_provider.py:
```diff
@@ -245,6 +264,14 @@ def check_3pid_auth(self, medium, address, password):
             )
             localpart = localpart[0] if len(localpart) == 1 else None
+            if self.ldap_ad_forest and localpart:
+                [login, domain] = localpart.rsplit("@", 1)
```
```suggestion
                (login, domain) = localpart.rsplit("@", 1)
```

richvdh commented on README.rst:
```diff
@@ -80,6 +80,50 @@ in search mode is provided below:
         bind_password: "ch33kym0nk3y"
         #filter: "(objectClass=posixAccount)"
 
```
I suggest a new section here:

```suggestion

Active Directory forest support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

richvdh commented on README.rst:
```diff
@@ -80,6 +80,50 @@ in search mode is provided below:
+separator (`=` or `/`) in the following form: `login<separator>domain`.
+
+Please note that common domain separators `@` and `\\` can not be used in Matrix
+User Indentifiers (https://matrix.org/docs/spec/appendices#user-identifiers)
```
Indentifiers -> Identifiers

richvdh commented on README.rst:
```diff
@@ -80,6 +80,50 @@ in search mode is provided below:
+        # Optional. Users from this domain may login
+        # without specifying domain part
+        default_domain: main.example.com
+        # Optional. Default /
+        mxid_domain_separator: /
```
```suggestion
        # Optional. Specifies the domain separator character that users should enter. Defaults to `/`.
        mxid_domain_separator: /
```

richvdh commented on README.rst:
```diff
@@ -80,6 +80,50 @@ in search mode is provided below:
         bind_password: "ch33kym0nk3y"
         #filter: "(objectClass=posixAccount)"
 
+Active Directory forest accounts are also supported with custom
+separator (`=` or `/`) in the following form: `login<separator>domain`.
+
+Please note that common domain separators `@` and `\\` can not be used in Matrix
+User Indentifiers (https://matrix.org/docs/spec/appendices#user-identifiers)
+and `/` separator is used by default.
+
+Let's say you have several domains in example.com forest:
```
I think this could be reworded a bit.

```suggestion
If the ``ad_forest`` flag is set to ``true``, an Active Directory forest will be
searched for the login details.

In this mode, the user enters their login details in the form ``<login>/<domain>``. (The 
normal domain separators, ``@`` and ``\\``, cannot be used in Matrix User Identifiers, so 
``/`` is used by default instead.)
```

richvdh commented on README.rst:
```diff
@@ -80,6 +80,50 @@ in search mode is provided below:
+        bind_password: "ch33kym0nk3y"
+        filter: "(objectClass=user)"
+
+With this configuration you can use logins:
```
```suggestion
With this configuration the user can log in with either ``someuser/main.example.com`` or ``someuser``.
```

etc

richvdh commented:
something else I forgot to say: we'll need sign-off from you to confirm that you are happy to contribute this code under the apache license. Sorry it's not spelt out more clearly, but please could you add a `Signed-off-by` to this PR as per https://github.com/matrix-org/synapse/blob/develop/CONTRIBUTING.md#sign-off ?

nE0sIghT commented:
> why not allow the user to enter <user>\<domain> in that form, and then map that to the matrix user ID @<user>/<domain>:<server_name> ?

Thanks for suggestion. I thinked about it but missed the fact that authentication plugin may return any mxid.

nE0sIghT commented:
> why not allow the user to enter <domain> in that form, and then map that to the matrix user ID @/:<server_name> ?

I reviewed [api documentation](https://github.com/matrix-org/synapse/blob/83434df3812650f53c60e91fb23c2079db0fb5b8/docs/password_auth_providers.md) and [Synapse source code](https://github.com/matrix-org/synapse/blob/83434df3812650f53c60e91fb23c2079db0fb5b8/synapse/handlers/auth.py#L723) and found that this is impossible with ```check_password``` method.  
Do you think we should switch to ```get_supported_login_types```/```check_auth``` pair from ```check_password```?

richvdh commented:
sorry for being so slow to reply - this had slipped off my radar.

>  Do you think we should switch to get_supported_login_types/check_auth pair from check_password?

I think that might be a good idea, yes.

richvdh commented:
please could you not force-push to git? it makes it much harder for us to see what has changed since the review.

nE0sIghT commented:
Ah, sorry for that.  
Do you want me to restore 1st commit and add changes as new commit?  

I still don't fixed tests, will do it today.

richvdh commented:
>  Do you want me to restore 1st commit and add changes as new commit?

that would be helpful, thanks!

nE0sIghT commented:
> that would be helpful, thanks!

Done. I restored original commit, but added ```Signed-off-by``` tag.  
I will squash commits after review if needed.

Thanks for your review!

richvdh commented on ldap_auth_provider.py:
```diff
@@ -76,18 +80,37 @@ def __init__(self, config, account_handler):
+                Canonical user ID if authentication against LDAP was successful
         """
+        password = login_dict['password']
         if not password:
             defer.returnValue(False)
```
if password is empty, then it should be handled the same way as any other incorrect password?

nE0sIghT replied:
Sounds reasonable :-)

richvdh commented on ldap_auth_provider.py:
```diff
@@ -48,6 +48,7 @@
 
 
 class ActiveDirectoryUPNException(Exception):
+    """Raises in case Matrix ID can not be mapped to UPN"""
```
it's not a matrix id.

```suggestion
    """Raised in case the user's login credentials cannot be mapped to a UPN"""
```

richvdh commented on ldap_auth_provider.py:
```diff
@@ -570,29 +578,43 @@ def _ldap_authenticated_search(self, server, password, filters):
             if not self.ldap_default_domain:
-                logger.debug(
+                logger.info(
                     'No LDAP separator %s was found in uid %s '
```
```suggestion
                    'No LDAP separator "/" was found in uid "%s" '
```

richvdh commented on ldap_auth_provider.py:
```diff
@@ -570,29 +578,43 @@ def _ldap_authenticated_search(self, server, password, filters):
                     'No LDAP separator %s was found in uid %s '
                     'and LDAP default domain was not configured. ',
-                    self.ldap_separator,
+                    "/",
```
```suggestion
```

richvdh commented on ldap_auth_provider.py:
```diff
@@ -570,29 +578,43 @@ def _ldap_authenticated_search(self, server, password, filters):
+                    "/",
                     self.ldap_default_domain
                 )
                 raise ActiveDirectoryUPNException()
```
seems like it might be simpler to just `return None` here, but :man_shrugging:

nE0sIghT replied:
If we will use ```None``` for that, then we can not just deserialize method output like

```
(login, domain, localpart) = self._map_login_to_upn(username)
```

So we will need some temporal variable like:
```
map_array = self._map_login_to_upn(username)
if not map_array:
    error here
(login, domain, localpart) = map_array
```

If you think it's better suits this project then I will refactor code. But I believe exception is better choice here.

richvdh replied:
fair enough.

richvdh commented on ldap_auth_provider.py:
```diff
@@ -347,6 +384,16 @@ class _LdapConfig(object):
 
+        ldap_config.active_directory = config.get("active_directory", False)
+        if ldap_config.active_directory:
+            if config["attributes"]["uid"].lower() != "userprincipalname":
+                raise Exception(
+                    "Only userPrincipalName is supported "
+                    "as uid property in active_directory mode"
+                )
```
ok, but it seems like this ought to be handled via documentation, rather than a hardcoded check. people may have systems that behave similarly to AD but expose a UPN-like attribute with a different name: there's no reason not to support that.

richvdh reviewed (changes requested):
getting there, thank you! a few more comments

nE0sIghT commented on ldap_auth_provider.py:
```diff
@@ -245,6 +271,17 @@ def check_3pid_auth(self, medium, address, password):
             )
             localpart = localpart[0] if len(localpart) == 1 else None
+            if self.ldap_active_directory and localpart:
+                (login, domain) = localpart.lower().rsplit("@", 1)
```
Added "@" check since we dropped ```userPrincipalName``` restriction.

nE0sIghT commented:
I added more commits. Hope they will address remaining issues.
Thanks!

richvdh commented on README.rst:
```diff
@@ -80,6 +80,52 @@ in search mode is provided below:
+        # Optional. Users from this domain may log in without specifying the domain part
+        default_domain: main.example.com
+        attributes:
+           # This must be set to `userPrincipalName` when `active_directory` is true
```
this is now incorrect (or at least misleading)?

richvdh commented on README.rst:
```diff
@@ -80,6 +80,52 @@ in search mode is provided below:
+Users of other domains in the ``example.com`` forest can log in with ``domain\login``
+or ``login/domain``.
+
+Please note that ``userPrincipalName`` or similary looking LDAP attribute in the format
```
```suggestion
Please note that ``userPrincipalName`` or a similar-looking LDAP attribute in the format
```

richvdh commented on README.rst:
```diff
@@ -80,6 +80,52 @@ in search mode is provided below:
+or ``login/domain``.
+
+Please note that ``userPrincipalName`` or similary looking LDAP attribute in the format
+``login@domain`` must be used when ``active_directory`` option is enabled.
```
```suggestion
``login@domain`` must be used when the ``active_directory`` option is enabled.
```

richvdh commented on ldap_auth_provider.py:
```diff
@@ -130,6 +128,10 @@ def check_auth(self, username, login_type, login_dict):
             )
 
             if self.ldap_mode == LDAPMode.SIMPLE:
+                if not password:
+                    # Password is mandatory in simple bind
+                    defer.returnValue(False)
```
I'm still a bit surprised by this. why do we need to special-case it? why is an empty password any different to any other incorrect password?

nE0sIghT replied:
That's how we could introduce security issue but thanks to ldap3 we don't.  
Long story short LDAP can not authenticate user without password. Attempting to bind user without password leads to anonymous bind which always success.  
Fortunately in ldap3 anonymous bind was splitted off to special case.  

I will revert empty password changes.

nE0sIghT commented:
I added more commits for remaining issues and added tests for new functionality as well.  
I also fixed connection leak in ```check_3pid_auth``` which prevents doing tests with that method.

richvdh reviewed (approved):
lgtm. Thanks very much for all your work!
